Remove getScriptRemoteCallType

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* **#57854 Remove getScriptRemoteCallType**
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

lw committed:
Remove getScriptRemoteCallType

Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -228,12 +222,9 @@ void RequestCallbackNoPython::processBaseScriptRemoteCall(
                               .toRRef());
   } else {
-    ownerRRef = ctx.getOrCreateOwnerRRef(rrefId, returnType);
+    ownerRRef = ctx.getOrCreateOwnerRRef(rrefId, jitFuture->elementType());
```
I assume `runJitFunction` and `runJitOperator` can retrieve type correctly, so that we can directly the the `elementType` from this Future?

lw replied:
I did look into how they determine the type for the Future that they return, and they seem to be doing some kind of introspection and type inference. Hence to me they seem to be doing the right thing, but I'm not an expert on JIT so I may perhaps miss some nuances?

mrshenli reviewed (approved):
LGTM

lw commented:
There turned out to be test failures because the inferred types were not "too" precise. In particular it seems before we were just inferring a "base" tensor whereas now we infer a "specialized" tensor. I had to thus relax a check later on to accept tensor subclasses.

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]

lw committed:
Update on "Remove getScriptRemoteCallType"


Because OwnerRRefs used to be created before their value was computed, we had to figure out their type ahead of time. After the previous diff, we inverted the order of operations, and we can now first compute the result and then create the OwnerRRef. Which means we can just inspect the value to get its type. Much simpler, and much less likely to get it wrong.

Differential Revision: [D28253843](https://our.internmc.facebook.com/intern/diff/D28253843/)

[ghstack-poisoned]
